File: dsa-problems/Greedy/solutions/cpp/GRD-008-exam-proctor-allocation.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

class Solution {
public:
    int minProctors(int n, int r, vector<pair<int,int>>& exams) {
        vector<pair<int,int>> events;
        events.reserve(2 * n);
        
        for (const auto& exam : exams) {
            events.push_back({exam.first, 1});  // Start
            events.push_back({exam.second, -1}); // End
        }
        
        // Sort: Time ascending. If times equal, Start (+1) before End (-1).
        // Since -1 < 1, default sort puts End before Start.
        // We want Start before End. So we need custom comparator.
        sort(events.begin(), events.end(), [](const pair<int,int>& a, const pair<int,int>& b) {
            if (a.first != b.first) return a.first < b.first;
            return a.second > b.second; // 1 > -1, so Start comes first
        });
        
        int maxOverlap = 0;
        int currentOverlap = 0;
        
        for (const auto& event : events) {
            currentOverlap += event.second;
            maxOverlap = max(maxOverlap, currentOverlap);
        }
        
        return (maxOverlap + r - 1) / r;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    
    int n, r;
    if (!(cin >> n >> r)) return 0;
    
    vector<pair<int,int>> exams(n);
    for (int i = 0; i < n; i++) {
        cin >> exams[i].first >> exams[i].second;
    }
    
    Solution solution;
    cout << solution.minProctors(n, r, exams) << "\n";
    
    return 0;
}

File: dsa-problems/Greedy/solutions/python/GRD-008-exam-proctor-allocation.py
import sys
import math

def min_proctors(n: int, r: int, exams: list) -> int:
    events = []
    for start, end in exams:
        events.append((start, 1))
        events.append((end, -1))
        
    # Sort events.
    # Primary key: time.
    # Secondary key: type. We want +1 before -1 to maximize overlap at boundaries.
    # In Python sort, (time, -1) comes before (time, 1).
    # So we should use type values that sort correctly, or reverse secondary sort.
    # Let's use type: 1 for start, -1 for end.
    # We want 1 before -1.
    # So sort by (time, -type) or custom key.
    events.sort(key=lambda x: (x[0], -x[1]))
    
    max_overlap = 0
    current_overlap = 0
    
    for _, type_ in events:
        current_overlap += type_
        max_overlap = max(max_overlap, current_overlap)
        
    # Ceiling division
    return (max_overlap + r - 1) // r

def main():
    input = sys.stdin.read
    data = input().split()
    if not data:
        return
        
    iterator = iter(data)
    n = int(next(iterator))
    r = int(next(iterator))
    
    exams = []
    for _ in range(n):
        start = int(next(iterator))
        end = int(next(iterator))
        exams.append([start, end])
    
    result = min_proctors(n, r, exams)
    print(result)

if __name__ == "__main__":
    main()
